addons/product/models/product.py

```diff
@@ -357,7 +357,8 @@ def unlink(self):
             if not product.exists():
                 continue
             # Check if the product is last product of this template...
-            other_products = self.search([('product_tmpl_id', '=', product.product_tmpl_id.id), ('id', '!=', product.id)])
+            other_products = self.env['product.product'].with_context(active_test=False).\
+                                    search([('product_tmpl_id', '=', product.product_tmpl_id.id), ('id', 'not in', self.ids)])
```

Comment:
This is not correct, we are iterating on products so you cannot use `self.ids` here - the search should be specific to each product, not to the whole set. It should use the leaf `('id', '!=', product.id)`. Otherwise, if you are deleting variants of the same template, they will interfere with each other.

Since we are looping on the selected products here, the last one to get deleted will mark the template as 'to be deleted' as well.